action/protocol/staking/protocol.go

```diff
@@ -468,6 +468,55 @@ func (p *Protocol) Validate(ctx context.Context, act action.Action, sr protocol.
 	return nil
 }
 
+func (p *Protocol) isActiveCandidate(ctx context.Context, csr CandidateStateReader, cand *Candidate) (bool, error) {
+	// at least min self stake
+	if cand.SelfStake.Cmp(p.config.RegistrationConsts.MinSelfStake) < 0 {
+		return false, nil
+	}
+
+	// endorsement must not exipred if the self-stake bucket is an endorse bucket
+	featureCtx := protocol.MustGetFeatureCtx(ctx)
+	if featureCtx.DisableDelegateEndorsement {
+		return true, nil
+	}
+	srHeight, err := csr.SR().Height()
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get StateReader height")
+	}
+	vb, err := csr.getBucket(cand.SelfStakeBucketIdx)
+	if err != nil {
+		if errors.Is(err, state.ErrStateNotExist) {
+			return false, nil
+		}
+		return false, err
+	}
+	// bucket is self-owned
+	if address.Equal(vb.Owner, cand.Owner) {
+		return true, nil
+	}
+	// bucket is not endorsed to the candidate
+	if !address.Equal(vb.Candidate, cand.Owner) {
+		return false, nil
```

Comment:
if this happens, there must be something wrong with our implementation?

Reply:
indeed, if clearing candidates is implemented after bucket endorsed to others